### ISSUE TITLE
Optimize Vert.x startup by avoiding DNS resolver initialization

### ISSUE THREAD

**Issue by gsmet** [quarkusio/quarkus#6678]
If we do something about this, it should be done in Vert.x but I'd like the discussion to start here to be sure it's really something that should be fixed in Vert.x or if it's something we should live with.

When starting Vert.x in `io/vertx/core/http/impl/HttpServerImpl.listen`, we resolve the address in `io/vertx/core/net/impl/AsyncResolveConnectHelper.doBind` via `io/vertx/core/impl/VertxImpl.resolveAddress`.

Most of the time (if not all), we will pass an IPv4/IPv6 address and I'm not sure we really need to resolve the name? Especially when using the default `0.0.0.0`.

From what I can see, with the default `0.0.0.0`, we set up the whole Vert.x DNS infrastructure at that point.

I have a few questions:
- do we need to do that for an IP address? It seems useless?
- if we don't, will the DNS infrastructure be set up later? What are the operations requiring address resolution? Will a RESTEasy request require DNS resolution?

Basically, is the effort to drop this worth it?

Note: I did some bootstrap profiling as my demo started to lag a lot in dev mode.

/cc @vietj @cescoffier @stuartwdouglas

**Comment by vietj** [comment]
the code will use a quick path in the resolver, 0.0.0.0 will be resolved quickly to an `InetSocketAddress` representing this address and not actual DNS resolution will happen (i.e no network call). Then this address will be used to do a Netty bind on it.

Not sure what is the issue here, unless you want to add a specific fast path that resolves hardcode the resolution of the string `0.0.0.0` to the `InetSocketAddress` and not call Netty.

**Comment by gsmet** [comment]
@vietj so... most of the bootstrap slowness is related to class loading. It's what really slows us down.

In this case, it's mostly related to the initialization of `NioDatagramChannel` and allegates. AFAICS this class is only called when initializing the DnsResolver layer (Eclipse says so but it might not be true, please correct me if incorrect).

If we don't resolve any other address later on, that would be quite beneficial to avoid loading all this machinery.

Now, I'm not sure if we do or don't. You're the expert here :).

**Comment by vietj** [comment]
have you tried in Quarkus native to turn JDK DNS resolver using `vertx.disableDnsResolver` system property, this should use the JDK dns resolver which will bypass this machinery ?

**Comment by gsmet** [comment]
I'm working on optimizing the dev mode experience so in JVM mode. Is your proposal valid for JVM mode, too?

I suppose you had your reasons to implement your own DNS resolver?

**Comment by cescoffier** [comment]
@vietj In native mode the Vert.x DNS resolver is disabled by default (users can opt-in).

**Comment by vietj** [comment]
why not then disable it by default in all cases ?

**Comment by gsmet** [comment]
I can try tomorrow and see how it goes.

**Comment by vietj** [comment]
the main reason is for performance reason when an application needs to constantly perform DNS resolution (like an HTTP gateway) then blocking DNS resolution can be a bottleneck.

**Comment by vietj** [comment]
however in case of a Quarkus app, the default to use the JDK resolver might be a better one.

**Comment by gsmet** [comment]
I created https://github.com/quarkusio/quarkus/pull/6706 . Looks like we had a small bug.

**Comment by cescoffier** [comment]
Initially, it was not working in native. SubstrateVM supports it now (I added an IT).

**Comment by gsmet** [comment]
What was not working in native? Should we remove the specific handling of native now and let the user decide which to use if both approaches work?

**Comment by cescoffier** [comment]
I don't remember exactly, but it was unable to resolve anything. It has been fixed in the recent GraalVM versions. 

I would disable the Vert.x async resolver by default. We still have issues in Kubernetes when using it. The user can opt-in if it works in his case.
